File: Workqueue/Mutex/C/mutex_benchmark.c
#include <linux/init.h>
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/mutex.h>
#include <linux/workqueue.h>
#include <linux/slab.h>
#include <linux/ktime.h>

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Luca Saverio Esposito");
MODULE_DESCRIPTION("Mutex Performance Test with Workqueue and Variable Increments");
MODULE_VERSION("1.0");

#define NUM_WORK_ITEMS 8
#define NUM_ITERATIONS 100000

struct shared_data {
    struct mutex lock;
    int value;
};

struct work_data {
    struct work_struct work;
    struct shared_data *shared;
    int id;
    ktime_t start_time;
    ktime_t end_time;
};

// Use the global system_long_wq workqueue
static struct workqueue_struct *test_wq = system_long_wq;

void simulate_delay(unsigned int delay)
{
    unsigned int i;
    volatile unsigned int temp = 0;

    for (i = 0; i < delay; i++)
        temp += i;
}

static void work_func(struct work_struct *work)
{
    struct work_data *data = container_of(work, struct work_data, work);
    int i, increment;

    data->start_time = ktime_get();

    // Determine increment based on thread ID
    if (data->id == 0)
        increment = 1;
    else if (data->id == 1)
        increment = 2;
    else
        increment = 3;

    for (i = 0; i < NUM_ITERATIONS; i++) {
        mutex_lock(&data->shared->lock);

        pr_info("Work item %d acquired mutex.\n", data->id);

        // Critical section: increment the shared variable
        data->shared->value += increment;

        pr_info("Work item %d incremented value to: %d \n", data->id, data->shared->value);

        mutex_unlock(&data->shared->lock);
        pr_info("Thread %d releasing mutex \n", data->id);
    }
    
    data->end_time = ktime_get();
    pr_info(    
        "Work item %d completed. \n Execution time: %lld ns\n",
        data->id,
        ktime_to_ns(ktime_sub(data->end_time, data->start_time))
    );
}

static int __init mutex_test_init(void)
{
    int i;
    ktime_t start_time, end_time;
    s64 elapsed_ns;

    pr_info("Initializing Mutex Performance Test with Workqueue...\n");

    // Allocate and initialize shared data
    struct shared_data *shared = kzalloc(sizeof(*shared), GFP_KERNEL);
    if (!shared)
        return -ENOMEM;

    mutex_init(&shared->lock);

    // Record start time
    start_time = ktime_get();

    // Initialize and queue work items
    for (i = 0; i < NUM_WORK_ITEMS; i++) {
        struct work_data *work_item = kzalloc(sizeof(*work_item), GFP_KERNEL);
        if (!work_item)
            continue;

        work_item->shared = shared;
        work_item->id = i;
        INIT_WORK(&work_item->work, work_func);

        queue_work(test_wq, &work_item->work);
    }

    // Record end time
    end_time = ktime_get();
    elapsed_ns = ktime_to_ns(ktime_sub(end_time, start_time));

    pr_info("Work items have been queued.\n");
    return 0;
}

static void __exit mutex_test_exit(void)
{
    pr_info("Exiting Mutex Performance Test with Workqueue.\n");
    pr_info("Final shared variable value: %d\n", shared->value);

    // Free shared data
    kfree(shared);
}

module_init(mutex_test_init);
module_exit(mutex_test_exit);

File: Workqueue/Mutex/Rust/mutex_benchmark.rs
// SPDX-License-Identifier: GPL-2.0

//! Test module for RwLock, Mutex, and Spinlock.

use kernel::{
    prelude::*,
    sync::{
        new_mutex,
        Arc,
        Mutex,
    },
    workqueue::{self, impl_has_work, new_work, Work, WorkItem},
};

module! {
    type: MutexTestModule,
    name: "mutex_test_module",
    author: "Luca Saverio Esposito",
    description: "Test module for Mutex",
    license: "GPL",
}

const NUM_WORK_ITEMS: u32 = 3;
const NUM_ITERATIONS: u32 = 1000; 


struct SharedData {
    value: u32,
}

struct MutexTestModule {
    mutex: Arc<Pin<Box<Mutex<SharedData>>>>,
}

impl kernel::Module for MutexTestModule {
    fn init(_module: &'static ThisModule) -> Result<Self> {
        pr_info!("MutexTestModule init\n");

        // Initialize the shared data structures and wrap them in Arc
        let mutex = Arc::new(Box::pin_init(
            new_mutex!(SharedData { value: 2 }), GFP_KERNEL
        )?,GFP_KERNEL)?;

        let module = Self {
            mutex,
        };

        // Start test threads
        module.test_mutex()?;

        Ok(module)
    }
}

impl MutexTestModule {
    fn test_mutex(&self) -> Result {
        pr_info!("Testing Mutex\n");

        let mutex = self.mutex.clone();

        // Create mutex work items
        for i in 0..NUM_WORK_ITEMS {
            let work_item = MutexWork::new(mutex.clone(), i)?;
            let _ = workqueue::system_long().enqueue(work_item); 
        }

        Ok(())
    }
}

impl Drop for MutexTestModule {
    fn drop(&mut self) {
        pr_info!("MutexTestModule exit\n");
    }
}

/// This function does dummy operation to simulate a delay
pub fn simulate_delay(delay: u32) {
    let mut a = 1;
    for _ in 0..delay {
        a = a + 1;
    }
}

#[pin_data]
struct MutexWork {
    thread_id: usize,
    #[pin]
    work: Work<MutexWork>,
    mutex: Arc<Pin<Box<Mutex<SharedData>>>>,
    start: Ktime,
    end: Ktime,
}

impl_has_work! {
    impl HasWork<Self> for MutexWork { self.work }
}

impl MutexWork {
    fn new(
        mutex: Arc<Pin<Box<Mutex<SharedData>>>>,
        thread_id: usize,
    ) -> Result<Arc<Self>> {
        Arc::pin_init(
            pin_init!(MutexWork {
                thread_id,
                mutex,
                work <- new_work!("MutexWork::work"),
            }),
            GFP_KERNEL,
        )
    }
}

impl WorkItem for MutexWork {
    type Pointer = Arc<Self>;

    fn run(this: Arc<Self>) {

        this.start = Ktime::ktime_get();

        // Simulate some operation
        if this.thread_id == 0 {
            guard.value += 1;
        } else if this.thread_id == 1 {
            guard.value += 2;
        } else {
            guard.value += 3;
        
        }

        for _ in  0..NUM_ITERATIONS{

            let mut guard = this.mutex.lock();
            
            pr_info!("Work item {} acquired mutex\n", this.thread_id);
            
            the_value = guard.value;
            pr_info!("Work item {} incremented value to: {}\n", this.thread_id, the_value,);

            pr_info!("Work item {} releasing mutex\n", this.thread_id);
            // Guard is dropped here
        }
        this.end = Ktime::ktime_get();
        pr_info!("Work item {} completed \n. Execution time: {} ms", this.thread_id, ktime_ms_delta(this.end,this.start));

    }
}
